### DISCUSSION TITLE
How to retrieve a secret where the secret variable name is stored inside another variable ?

### DISCUSSION THREAD

**Discussion by ppmgitrepo** [community/community#116494] [Actions] [answered]
### Select Topic Area

Not able to derive a Secret value , when the name of the Secret variable is dynamic and not static. 

### Body

example 
Against an Environment multiple secrets would be defined 
DB_APP1_PASSWORD
DB_APP2_PASSWORD 
DB_APP3_PASSWORD 
etc..

Now at runtime I will derive which secret to be used and that secret variable name will be built as a string inside another variable
 
v_secret='DB_APP1_PASSWORD'
or depending on the context it can be 
v_secret='DB_APP2_PASSWORD'

However this is not getting resolved at runtime i.e. I tried various methods of binding.

```
    run: |
      v_secret='DB_APP1_PASSWORD'
      vDB_PASSWORD=${{ secrets[v_secret] }}  # Access secret using constructed expression
      echo "The database password retrieved dynamically is $vDB_PASSWORD"
```
      Note: I understand it will not echo, but the binding itself is not resolving. 
      Any help in the syntax to resolve this would be much appreciated.

**Comment by Gumaan-Mahar** [comment]
In GitHub Actions, you cannot directly reference dynamic variable names like that. Instead, you need to use `env` context and evaluate the dynamic variable name before accessing the secret. Here's how you can do it:

```yaml
name: Retrieve Secret

on:
  push:
    branches:
      - main

jobs:
  retrieve_secret:
    runs-on: ubuntu-latest
    steps:
      - name: Set dynamic secret name
        run: echo "SECRET_NAME=DB_APP1_PASSWORD" >> $GITHUB_ENV

      - name: Retrieve secret
        run: |
          v_secret=$SECRET_NAME
          vDB_PASSWORD="${{ secrets[v_secret] }}"
          echo "The database password retrieved dynamically is $vDB_PASSWORD"
```

- The `Set dynamic secret name` step sets the dynamic secret name as an environment variable.
- In the `Retrieve secret` step, we first read the value of `SECRET_NAME` from the environment, then use that to construct the name of the secret variable (`v_secret`). Finally, we use `${{ secrets[v_secret] }}` to retrieve the secret's value.

Make sure to replace `DB_APP1_PASSWORD` with your actual secret names accordingly. This way, you can dynamically fetch secrets based on runtime variables in GitHub Actions.

**Reply by ppmgitrepo** [reply]
@Gumaan-Mahar 

Thanks for your response. I tried it however it does not work. 
i.e. before it runs itself it fails at this line 

`
vDB_PASSWORD="${{ secrets[v_secret] }}"

`

Error : Unrecognized named-value: 'v_secret'. Located at position 9 within expression: secrets[v_secret]

**Reply by Gumaan-Mahar** [reply, marked as answer]
My bad! Let's use the **env** directive to set the environment variable first, and then access it directly. Here's the updated code:

```
- name: Retrieve secret
  env:
    SECRET_NAME: ${{ env.SECRET_NAME }}
  run: |
    vDB_PASSWORD="${{ secrets[env.SECRET_NAME] }}"
    echo "The database password retrieved dynamically is $vDB_PASSWORD"
```

Just give me a heads up if it's still not working, okay?

**Reply by ppmgitrepo** [reply]
I deleted my previous comment. 

This method works and the binding is happing.
Thanks a lot for the solution. I have been struggling with this for a couple of days 
Appreciate the help :-D

**Reply by ppmgitrepo** [reply]
@Gumaan-Mahar 

I have one more question , the problem is I have a while loop that processes many rows. 
and within the while loop the SECRET_NAME will need to be switched. 

But it looks like there is no way to implement this because within a job it appears that its not possible to switch the value of an env variable

i.e.
```
  env:
    SECRET_NAME: ${{ env.SECRET_NAME }}
```

Is there any way you can re-set the value of env.SECRET_NAME within the same job ?
I tried this, but it does not work , it still retrieves the old value

```
          echo "SECRET_NAME=DB_APP2_PASSWORD" >> $GITHUB_ENV
         ## and then retrive the value 
          vDB_PASSWORD="${{ secrets[env.SECRET_NAME ] }}"
```

Because within the While loop different rows will need to connect to different db schemas hence the SECRET needs to be switched.

**Reply by kingthorin** [reply]
@ppmgitrepo did you ever come up with a solution for this?
I have a similar requirement. ex: I have a set of values (say in a flat file), those values will be the prefix of the secret I need to access. I want to be able to loop in a shell script or run block and work with the secrets.

**Reply by RawRapter** [reply]
@ppmgitrepo Hey facing the same issue, I want to dynamically fetch the secret into a variable that, then uses that variable for the secret.
I can dynamically get the secret variable key and pass it to another environment variable but I cannot get the secret with that new variable.

**Reply by kingthorin** [reply]
Instead of trying to generate and access dynamic secret names we decided we could put a whole set of secrets into a single GitHub secret and load them into the docker env, that made things much more workable for us.

GitHub secrets can be up to 48k, which can be quite a few name value pairs.

**Comment by learnprofile** [comment]
@ppmgitrepo @Gumaan-Mahar , I am trying your same approach as i am trying to pull all my environment secrets but it gives me same error.

 i just tried by creating entire new workflow but i am getting this error
Unrecognized named-value: 'SECRET_NAME'. Located at position 9 within expression: secrets[SECRET_NAME]


name: Dynamic Secrets Workflow

on:
workflow_dispatch:

push:
branches:
- main

jobs:
fetch_secrets:
runs-on: ubuntu-latest
outputs:
secret_names_string: ${{ steps.get_secrets.outputs.secret_names_string }}
steps:
- name: Checkout code
uses: actions/checkout@v4

  - name: Get Secret Names
    id: get_secrets
    env:
      GITHUB_TOKEN: ${{ secrets.SOURCE_PAT }}
      ENVIRONMENT_NAME: test  
    run: |
      SECRETS_JSON=$(gh api repos/${{ github.repository }}/environments/test/secrets --jq '[.secrets[].name]')
      SECRET_NAMES_STRING=$(echo $SECRETS_JSON | jq -r '. | join(" ")')

      echo "secret_names_string=$SECRET_NAMES_STRING" >> $GITHUB_OUTPUT
use_secrets:
needs: fetch_secrets
runs-on: ubuntu-latest
env:
SECRET_NAMES_STRING: ${{ needs.fetch_secrets.outputs.secret_names_string }}
steps:
- name: Checkout code
uses: actions/checkout@v4

  - name: Use Secrets Dynamically
    run: |
      SECRET_NAMES=($SECRET_NAMES_STRING)
      for SECRET_NAME in "${SECRET_NAMES[@]}"; do
        SECRET_VALUE="${{ secrets[SECRET_NAME] }}"
        echo "Secret Name: $SECRET_NAME"
        echo "Secret Value: $SECRET_VALUE"
        echo "::set-env name=$SECRET_NAME::$SECRET_VALUE"
      done
at this line : SECRET_NAMES=($SECRET_NAMES_STRING) shows
Unrecognized named-value: 'SECRET_NAME'. Located at position 9 within expression: secrets[SECRET_NAME]

am i missing anything?

**Comment by learnprofile** [comment]
@ppmgitrepo , @kingthorin @RawRapter @Gumaan-Mahar , Do you have any inputs for me to try anything else?

**Reply by kingthorin** [reply]
Your workflow seems to mix get-secrets and fetch-secrets

**Comment by learnprofile** [comment]
Thats right my intention is to fetch secret key names first and then create a json body to fetch secrets as a part of jq command as currently to fetch secrets in base 64 i have to manually add secret names in github workflow.

**Reply by kingthorin** [reply]
Your needs directive was throwing me off. I don't see what else might be wrong but it's hard to follow without proper formatting.

**Reply by learnprofile** [reply]
Let me know if this works for you and see if you can trace any formatting issue
